File: 算法/Python/第3章 排序算法/3.6 快速排序/quick_sort.py
def quick_sort(list, start, end):
    """
        快速排序
        Args:
            list (list): 待排序列表
            start (int): 开始下标
            end (int): 结束下标
    """
    if start < end:
        i = start
        j = end
        pivot = list[start]

        while i < j:
            while i < j and list[j] > pivot:
                j -= 1
            if i < j:
                list[i] = list[j]
                i += 1
            while i < j and list[i] < pivot:
                i += 1
            if i < j:
                list[j] = list[i]
                j -= 1
        
        list[i] = pivot
        quick_sort(list, start, i-1)
        quick_sort(list, i+1, end)

def main():
    list = [5, 8, 6, 3, 9, 2, 1, 7]
    print("排序前：" + str(list))
    quick_sort(list, 0, len(list)-1)
    print("排序后：" + str(list))

if __name__ == "__main__":
    main()

File: code/Java/第8章 排序算法/8.6 快速排序/QuickSort.java
import java.util.Arrays;

public class QuickSort {
	public static void main(String[] args) {
		int[] arr = {5, 8, 6, 3, 9, 2, 1, 7};
		System.out.print("排序前：");
	    System.out.println(Arrays.toString(arr));
	    quickSort(arr, 0, arr.length-1);
	    System.out.print("排序后：");
	    System.out.println(Arrays.toString(arr));
	}
	
	/**
	 * @brief  快速排序
	 * @param  arr: 待排序数组
	 * @param  start: 开始下标
	 * @param  end: 结束下标
	 * @retval None
	 */
	public static void quickSort(int[] arr, int start, int end) {
	    if(start < end) {
	        int i = start;
	        int j = end;
	        int pivot = arr[start];

	        while(i < j) {
	            while(i < j && arr[j] > pivot) {
	                j--;
	            }
	            if(i < j) {
	                arr[i] = arr[j];
	                i++;
	            }
	            while(i < j && arr[i] < pivot) {
	                i++;
	            }
	            if(i < j) {
	                arr[j] = arr[i];
	                j--;
	            }
	        }
	        arr[i] = pivot;
	        quickSort(arr, start, i-1);
	        quickSort(arr, i+1, end);
	    } 
	}
}